File: server/src/pricing.js
export function currentMultiplier(occupancy) {
  if (occupancy < 0.5) return 1.0;
  if (occupancy < 0.8) return 1.2;
  return 1.5;
}

function timeBandMultiplier(now) {
  const t = now.getHours() + now.getMinutes() / 60;
  const inMorning = t >= 7 && t <= 10;
  const inEvening = t >= 16 && t <= 19.5;
  if (inMorning || inEvening) return 1.15;
  return 1.0;
}

export function computePrices(grid) {
  const now = new Date();
  const lotOcc = grid.occupancy();
  const base = 2.5;
  const surge = currentMultiplier(lotOcc) * timeBandMultiplier(now);
  const prices = {};
  grid.slots.forEach((slot, id) => {
    const zoneOcc = grid.occupancy(slot.zone);
    const zoneSurge = currentMultiplier(zoneOcc);
    const locationFactor = slot.zone === "A" ? 1.1 : 1.0;
    prices[id] = Number((base * surge * zoneSurge * locationFactor).toFixed(2));
  });
  return prices;
}

File: backend/app/pricing.py
from __future__ import annotations

from datetime import datetime, time
from typing import Dict

from .state import GridState


def current_multiplier(occupancy: float) -> float:
    if occupancy < 0.5:
        return 1.0
    if occupancy < 0.8:
        return 1.2
    return 1.5


def time_band_multiplier(now: datetime) -> float:
    morning_peak = time(7, 0), time(10, 0)
    evening_peak = time(16, 0), time(19, 30)
    t = now.time()
    if morning_peak[0] <= t <= morning_peak[1] or evening_peak[0] <= t <= evening_peak[1]:
        return 1.15
    return 1.0


def compute_prices(grid: GridState) -> Dict[str, float]:
    """Return suggested price per slot id."""
    now = datetime.utcnow()
    lot_occ = grid.occupancy()
    base = 2.5
    surge = current_multiplier(lot_occ) * time_band_multiplier(now)
    prices: Dict[str, float] = {}
    for slot_id, slot in grid.slots.items():
        zone_occ = grid.occupancy(slot.zone)
        zone_surge = current_multiplier(zone_occ)
        location_factor = 1.1 if slot.zone == "A" else 1.0
        prices[slot_id] = round(base * surge * zone_surge * location_factor, 2)
    return prices

